### ISSUE TITLE
Busy Disks Cause Lag

### ISSUE THREAD

**Issue by Warr1024** [minetest/minetest#15125]
### Minetest version

```true
Minetest 5.9.0 (Linux)
Using LuaJIT 2.1.0-beta3
Built by GCC 12.2
Running on Linux/6.1.0 x86_64
BUILD_TYPE=RelWithDebInfo
RUN_IN_PLACE=1
USE_CURL=1
USE_GETTEXT=1
USE_SOUND=1
STATIC_SHAREDIR="."
STATIC_LOCALEDIR="locale"
```

### Irrlicht device

N/A (happens on both SP and dedicated servers)

### Operating system and version

Debian 12, various Alpine versions, etc.

### CPU model

Various (Intel i5-3320M (4) @ 3.300GHz, AMD EPYC 7282 (4) @ 2.794GHz, Intel i5-2400 (4) @ 3.400GHz, etc)

### GPU model

N/A (happens on both SP and dedicated servers)

### Active renderer

N/A (happens on both SP and dedicated servers)

### Summary

Background disk activity causes main thread lag on MT servers.

It seems like this may be somewhat worse on some versions than others, but it's present in all versions.

This can significantly worsen the player experience when doing certain things like backups or database maintenance.

Given that MT is already very asynchronous and non-durable (things like map emerge APIs are all async, and writes like mod_storage are already deferred to a timer) there is no reason why the main game thread should ever be waiting on the disk.

### Steps to reproduce

- Install szutilpack from CDB and enable szutil_lagometer for a world.
- Load up the world.
- Run the /lagometer command to enable the lagometer.
- Fly around, loading up terrain.  Note the baseline lag profile.
- Quit the world and restart it (mapblocks are now unloaded again).
- Start a disk-intensive process in the background, like bonnie++.
- Fly around, loading up terrain.  Note the new lag profile.

### Observed

Lag goes through the roof.  You might have some main thread globalsteps that take several seconds to complete.

### Expected

Unloaded terrain remains unloaded for longer, as the disk takes time to respond, but globalstep continues in the meantime.  Lag should remain at roughly the same baseline level as with a quiescent disk.

**Comment by SmallJoker** [comment]
What's the hot path that causes this delay? Is it map saving or something else?

**Comment by appgurueu** [comment]
Can you try to [get a profile of the time spent sleeping](https://perf.wiki.kernel.org/index.php/Tutorial#Profiling_sleep_times)?

**Comment by Warr1024** [comment]
I was able to record, but after running "inject", then "report" would throw a "data has no samples!" error.  I tried running the same "report" line against the "raw" files, but I couldn't make sense of them.

Attaching the raw files for your examination:

- [baseline.data.raw.gz](https://github.com/user-attachments/files/16932729/baseline.data.raw.gz)
- [busydisk.data.raw.gz](https://github.com/user-attachments/files/16932730/busydisk.data.raw.gz)

**Comment by Desour** [comment]
Here, for testing (with annotated env lock, and a bunch of zonescopeds):
https://github.com/Desour/minetest/tree/tracy_profiling_pr_tmp2

I haven't tested myself with busy disk. But it's no surprise if causes server lag. Block loading from db, deserialization, and even reflow scan all happen under the env lock, even if the emerge thread does it.

**Comment by Warr1024** [comment]
> Block loading from db, deserialization, and even reflow scan all happen under the env lock, even if the emerge thread does it.

That's the impression I got from elsewhere, yeah.

That's ... pretty bad.  It goes a long way to defeat the purpose of even having a separate thread to begin with.

The emerge thread should only need to acquire the env lock _briefly_ to mark the block as being in an "I'm already working on it now" state, and then run all that stuff outside the lock, and then acquire the lock again just to "attach" the loaded block to the env and let it finish rehydrating things like entities and timers and such.

**Comment by swagtoy** [comment]
Just wanted to add but:

> Switching backends to postgresql

This might improve scaling probably, but it won't actually fix the issue at question (unless Postgresql has some asynchronous scaling nonsense)

**Comment by Warr1024** [comment]
Right, it's a partial mitigation.  Postgresql has some advantages that have little to do with performance.  It's a little better configured by default than sqlite, but if both are properly configured, their performance is comparable.  That means it's also possible to reproduce this issue on multiple backends, as long as the relevant disks are busy.

**Comment by sfan5** [comment]
> Block loading from db [...] happen[s] under the env lock, even if the emerge thread does it.

I tested this and expected to falsify this claim, but it's true.
This sounds like someone made a terrible mistake somewhere because this is 50% of why we have the emerge thread at all.

**Comment by Warr1024** [comment]
Right, this was never a ***new*** issue in 5.9, it was just ***present*** in 5.9.

I first noticed the fact that slow databases could cause lag probably back in the 5.3 days.  It's been an open secret for a while, but usually it wasn't a *big* effect.  At the time, I noticed about a half-second lagspike every 5 seconds, which was apparently due to sqlite writes.  Enabling WAL mode changed that to a ~1s spike every 60 seconds.  I moved to postgresql and the lagspikes due to _writes_ seemed to disappear, but I still noticed a lot of lag on _reads_.  My ultimate solution for that particular server was to upgrade to an SSD.

Now, a few years later, I've got more players, a more complex game, and more demanding users, and after a big "lag crisis" that coincided with updating my servers to 5.9, I've started paying attention to this stuff again.  I'm a little puzzled, too, why this hadn't been reported before, but :shrug: I guess every project needs ONE "worst under-reported bug".

It looks like it's not just the database that's affected, but some of the other files, like the ones going through "safe writes."  Having those instantly durable seems weird, given that MT delays everything else by ~5s.  Those should probably still be done atomically (so that the resultant files are consistent) but asynchronously as well.

Moving the database outside of the lock though is probably like a 95% to 99% fix or something, so I'm eager to test that out.

**Comment by kromka-chleba** [comment]
Lags upon map generation/loading can be easily seen with the [Legboat mod](https://content.minetest.net/packages/ekl/legboat/) in MTG. Every time a chunk emerges the legboat lags.

**Comment by Warr1024** [comment]
Not under load:

```
2024-09-11 16:17:02: INFO[Main]: Profiler:
2024-09-11 16:17:02: INFO[Main]:   ActiveObjectMgr: CAO count [#] ______________  128x       1
2024-09-11 16:17:02: INFO[Main]:   ActiveObjectMgr: SAO count [#] ______________  276x       1
2024-09-11 16:17:02: INFO[Main]:   CM::Transparent Buffers - Sorted ____________  139x   5.079
2024-09-11 16:17:02: INFO[Main]:   CM::Transparent Buffers - Unsorted __________  139x  620.345
2024-09-11 16:17:02: INFO[Main]:   CM::reportMetrics loaded blocks [#] _________    1x    5977
2024-09-11 16:17:02: INFO[Main]:   CM::touchMapBlocks() [ms] 
2024-09-11 16:17:02: INFO[Main]:   CM::updateDrawList() [ms] ___________________   26x   2.115
2024-09-11 16:17:02: INFO[Main]:   CM::updateDrawListShadow() [ms] _____________   17x    1.47
2024-09-11 16:17:02: INFO[Main]:   CM::updateTransparentMeshBuffers [ms] _______  139x   0.215
2024-09-11 16:17:02: INFO[Main]:   Client: Mesh making (sum) [ms] ______________    1x    1659
2024-09-11 16:17:02: INFO[Main]:   Client: collisionMoveSimple() [us] 
2024-09-11 16:17:02: INFO[Main]:   ClientEnv: CSO count [#] 
2024-09-11 16:17:02: INFO[Main]:   Clouds::updateMesh() [ms] ___________________    3x   0.333
2024-09-11 16:17:02: INFO[Main]:   EmergeThread: Lua on_generated [ms] 
2024-09-11 16:17:02: INFO[Main]:   EmergeThread: Mapgen::makeChunk [ms] 
2024-09-11 16:17:02: INFO[Main]:   EmergeThread: after Mapgen::makeChunk [ms] 
2024-09-11 16:17:02: INFO[Main]:   EmergeThread: load block - async (sum) [ms] _    1x      47
2024-09-11 16:17:02: INFO[Main]:   EmergeThread: update lighting [ms] 
2024-09-11 16:17:02: INFO[Main]:   Game::updateFrame(): update frame [ms] ______  139x  34.647
2024-09-11 16:17:02: INFO[Main]:   Irr: buffers uploaded _______________________  139x  157.057
2024-09-11 16:17:02: INFO[Main]:   Irr: buffers uploaded (bytes) _______________  139x  1.36993e+06
2024-09-11 16:17:02: INFO[Main]:   Irr: drawcalls ______________________________  139x  3358.96
2024-09-11 16:17:02: INFO[Main]:   Irr: primitives per drawcall ________________  139x  200.542
2024-09-11 16:17:02: INFO[Main]:   MapBlock meshes in range [#] ________________   26x  5939.35
2024-09-11 16:17:02: INFO[Main]:   MapBlocks drawn [#] _________________________   26x  699.923
2024-09-11 16:17:02: INFO[Main]:   MapBlocks examined [#] 
2024-09-11 16:17:02: INFO[Main]:   MapBlocks frustum culled [#] ________________   26x  4799.54
2024-09-11 16:17:02: INFO[Main]:   MapBlocks loaded [#] ________________________   26x  5939.35
2024-09-11 16:17:02: INFO[Main]:   MapBlocks occlusion culled [#] 
2024-09-11 16:17:02: INFO[Main]:   MapBlocks shortlist [#] 
2024-09-11 16:17:02: INFO[Main]:   MapBlocks sides skipped [#] 
2024-09-11 16:17:02: INFO[Main]:   ParticleManager: particle buffer count [#] 
2024-09-11 16:17:02: INFO[Main]:   SEnv: modify in blocks avg per interval [ms]     5x     9.8
2024-09-11 16:17:02: INFO[Main]:   SHADOW MapBlock meshes in range [#] _________   17x  412.47
2024-09-11 16:17:02: INFO[Main]:   SHADOW MapBlocks drawn [#] __________________   17x  412.47
2024-09-11 16:17:02: INFO[Main]:   SHADOW MapBlocks loaded [#] _________________   17x  5968.23
2024-09-11 16:17:02: INFO[Main]:   Server: Process network packet (sum) [ms] 
2024-09-11 16:17:02: INFO[Main]:   Server: liquid transform [ms] _______________    1x       8
2024-09-11 16:17:02: INFO[Main]:   Server: map saving (sum) [ms] _______________    1x      29
2024-09-11 16:17:02: INFO[Main]:   Server: map timer and unload [ms] ___________    1x      84
2024-09-11 16:17:02: INFO[Main]:   Server: send SAO messages [ms] 
2024-09-11 16:17:02: INFO[Main]:   Server: update objects within range [ms] 
2024-09-11 16:17:02: INFO[Main]:   Server::AsyncRunStep() [ms] _________________  276x   0.985
2024-09-11 16:17:02: INFO[Main]:   Server::RunStep() (max) [ms] ________________    1x      89
2024-09-11 16:17:02: INFO[Main]:   Server::SendBlocks(): Collect list [ms] _____    1x     315
2024-09-11 16:17:02: INFO[Main]:   Server::SendBlocks(): Send to clients [ms] __    1x      81
2024-09-11 16:17:02: INFO[Main]:   ServerEnv: ABMs run _________________________    5x    70.4
2024-09-11 16:17:02: INFO[Main]:   ServerEnv: Run SAO::step() [ms] _____________  276x   0.003
2024-09-11 16:17:02: INFO[Main]:   ServerEnv: Run node timers [ms] _____________   46x   0.326
2024-09-11 16:17:02: INFO[Main]:   ServerEnv: active blocks ____________________    5x   423.2
2024-09-11 16:17:02: INFO[Main]:   ServerEnv: active blocks cached _____________    5x  243.199
2024-09-11 16:17:02: INFO[Main]:   ServerEnv: active blocks scanned for ABMs ___    5x   147.8
2024-09-11 16:17:02: INFO[Main]:   ServerEnv: move players [ms] 
2024-09-11 16:17:02: INFO[Main]:   ServerEnv: update active blocks [ms] ________    3x  11.666
2024-09-11 16:17:02: INFO[Main]:   ServerEnv::step() [ms] ______________________  276x   0.467
2024-09-11 16:17:02: INFO[Main]:   ServerEnvironment::removeRemovedObjects() [ms] 
2024-09-11 16:17:02: INFO[Main]:   ServerMap: deSer block [us] _________________  725x   92.38
2024-09-11 16:17:02: INFO[Main]:   ServerMap: load block - sync (sum) [ms] 
2024-09-11 16:17:02: INFO[Main]:   ServerMap: load block [us] __________________  725x  149.555
2024-09-11 16:17:02: INFO[Main]:   Sky::render() [us] __________________________  139x   89.87
2024-09-11 16:17:02: INFO[Main]:   renderMap(): animated meshes [#] ____________  139x  35.489
2024-09-11 16:17:02: INFO[Main]:   renderMap(): transparent buffers [#] ________  139x  38.273
2024-09-11 16:17:02: INFO[Main]:   renderMap(SHADOW SOLID): draw meshes [ms] ___  136x   2.602
2024-09-11 16:17:02: INFO[Main]:   renderMap(SHADOW SOLID): drawcalls [#] ______  136x  378.727
2024-09-11 16:17:02: INFO[Main]:   renderMap(SHADOW SOLID): material swaps [#] _  136x   38.75
2024-09-11 16:17:02: INFO[Main]:   renderMap(SHADOW SOLID): vertices drawn [#] _  136x  168121
2024-09-11 16:17:02: INFO[Main]:   renderMap(SHADOW TRANS): draw meshes [ms] ___   17x   0.352
2024-09-11 16:17:02: INFO[Main]:   renderMap(SHADOW TRANS): drawcalls [#] ______   17x  47.176
2024-09-11 16:17:02: INFO[Main]:   renderMap(SHADOW TRANS): material swaps [#] _   17x  47.176
2024-09-11 16:17:02: INFO[Main]:   renderMap(SHADOW TRANS): vertices drawn [#] _   17x  22876.7
2024-09-11 16:17:02: INFO[Main]:   renderMap(SOLID): draw meshes [ms] __________  139x   21.05
2024-09-11 16:17:02: INFO[Main]:   renderMap(SOLID): drawcalls [#] _____________  139x  2870.99
2024-09-11 16:17:02: INFO[Main]:   renderMap(SOLID): material swaps [#] ________  139x  90.733
2024-09-11 16:17:02: INFO[Main]:   renderMap(SOLID): vertices drawn [#] ________  139x  1.14305e+06
2024-09-11 16:17:02: INFO[Main]:   renderMap(TRANSPARENT): draw meshes [ms] ____  139x   0.431
2024-09-11 16:17:02: INFO[Main]:   renderMap(TRANSPARENT): drawcalls [#] _______  139x  38.273
2024-09-11 16:17:02: INFO[Main]:   renderMap(TRANSPARENT): material swaps [#] __  139x  38.273
2024-09-11 16:17:02: INFO[Main]:   renderMap(TRANSPARENT): vertices drawn [#] __  139x  14017.1
```

Under load:

```
2024-09-11 16:19:48: INFO[Main]: Profiler:
2024-09-11 16:19:48: INFO[Main]:   ActiveObjectMgr: CAO count [#] ______________  627x       1
2024-09-11 16:19:48: INFO[Main]:   ActiveObjectMgr: SAO count [#] ______________  865x       1
2024-09-11 16:19:48: INFO[Main]:   CM::Transparent Buffers - Sorted ____________  639x   0.086
2024-09-11 16:19:48: INFO[Main]:   CM::Transparent Buffers - Unsorted __________  639x  397.198
2024-09-11 16:19:48: INFO[Main]:   CM::reportMetrics loaded blocks [#] _________    3x  4387.33
2024-09-11 16:19:48: INFO[Main]:   CM::touchMapBlocks() [ms] 
2024-09-11 16:19:48: INFO[Main]:   CM::updateDrawList() [ms] ___________________   95x   1.536
2024-09-11 16:19:48: INFO[Main]:   CM::updateDrawListShadow() [ms] _____________   78x   1.269
2024-09-11 16:19:48: INFO[Main]:   CM::updateTransparentMeshBuffers [ms] _______  639x   0.158
2024-09-11 16:19:48: INFO[Main]:   Client: Mesh making (sum) [ms] ______________    1x    3232
2024-09-11 16:19:48: INFO[Main]:   Client: collisionMoveSimple() [us] 
2024-09-11 16:19:48: INFO[Main]:   ClientEnv: CSO count [#] 
2024-09-11 16:19:48: INFO[Main]:   Clouds::updateMesh() [ms] 
2024-09-11 16:19:48: INFO[Main]:   EmergeThread: load block - async (sum) [ms] _    1x    1670
2024-09-11 16:19:48: INFO[Main]:   Game::updateFrame(): update frame [ms] ______  639x  24.851
2024-09-11 16:19:48: INFO[Main]:   Irr: buffers uploaded _______________________  639x  44.477
2024-09-11 16:19:48: INFO[Main]:   Irr: buffers uploaded (bytes) _______________  639x  252524
2024-09-11 16:19:48: INFO[Main]:   Irr: drawcalls ______________________________  639x  1696.21
2024-09-11 16:19:48: INFO[Main]:   Irr: primitives per drawcall ________________  639x  140.604
2024-09-11 16:19:48: INFO[Main]:   MapBlock meshes in range [#] ________________   95x  4386.72
2024-09-11 16:19:48: INFO[Main]:   MapBlocks drawn [#] _________________________   95x  458.684
2024-09-11 16:19:48: INFO[Main]:   MapBlocks examined [#] 
2024-09-11 16:19:48: INFO[Main]:   MapBlocks frustum culled [#] ________________   95x  3609.8
2024-09-11 16:19:48: INFO[Main]:   MapBlocks loaded [#] ________________________   95x  4386.72
2024-09-11 16:19:48: INFO[Main]:   MapBlocks occlusion culled [#] 
2024-09-11 16:19:48: INFO[Main]:   MapBlocks shortlist [#] 
2024-09-11 16:19:48: INFO[Main]:   MapBlocks sides skipped [#] 
2024-09-11 16:19:48: INFO[Main]:   ParticleManager: particle buffer count [#] 
2024-09-11 16:19:48: INFO[Main]:   SEnv: modify in blocks avg per interval [ms]    18x   7.666
2024-09-11 16:19:48: INFO[Main]:   SHADOW MapBlock meshes in range [#] _________   78x  303.32
2024-09-11 16:19:48: INFO[Main]:   SHADOW MapBlocks drawn [#] __________________   78x  303.32
2024-09-11 16:19:48: INFO[Main]:   SHADOW MapBlocks loaded [#] _________________   78x  4365.58
2024-09-11 16:19:48: INFO[Main]:   Server: Process network packet (sum) [ms] ___    1x       2
2024-09-11 16:19:48: INFO[Main]:   Server: liquid transform [ms] _______________    1x       1
2024-09-11 16:19:48: INFO[Main]:   Server: map saving (sum) [ms] _______________    1x      74
2024-09-11 16:19:48: INFO[Main]:   Server: map timer and unload [ms] ___________    1x     607
2024-09-11 16:19:48: INFO[Main]:   Server: send SAO messages [ms] ______________    1x       2
2024-09-11 16:19:48: INFO[Main]:   Server: update objects within range [ms] ____    1x       1
2024-09-11 16:19:48: INFO[Main]:   Server::AsyncRunStep() [ms] _________________  865x   1.394
2024-09-11 16:19:48: INFO[Main]:   Server::RunStep() (max) [ms] ________________    1x     366
2024-09-11 16:19:48: INFO[Main]:   Server::SendBlocks(): Collect list [ms] _____    1x     594
2024-09-11 16:19:48: INFO[Main]:   Server::SendBlocks(): Send to clients [ms] __    1x     196
2024-09-11 16:19:48: INFO[Main]:   ServerEnv: ABMs run _________________________   18x  16.222
2024-09-11 16:19:48: INFO[Main]:   ServerEnv: Run SAO::step() [ms] _____________  865x   0.009
2024-09-11 16:19:48: INFO[Main]:   ServerEnv: Run node timers [ms] _____________  184x   0.684
2024-09-11 16:19:48: INFO[Main]:   ServerEnv: active blocks ____________________   18x  460.333
2024-09-11 16:19:48: INFO[Main]:   ServerEnv: active blocks cached _____________   18x  300.888
2024-09-11 16:19:48: INFO[Main]:   ServerEnv: active blocks scanned for ABMs ___   18x  68.166
2024-09-11 16:19:48: INFO[Main]:   ServerEnv: move players [ms] 
2024-09-11 16:19:48: INFO[Main]:   ServerEnv: update active blocks [ms] ________    9x  10.777
2024-09-11 16:19:48: INFO[Main]:   ServerEnv::step() [ms] ______________________  865x   0.517
2024-09-11 16:19:48: INFO[Main]:   ServerEnvironment::removeRemovedObjects() [ms] 
2024-09-11 16:19:48: INFO[Main]:   ServerMap: deSer block [us] _________________ 2058x  96.224
2024-09-11 16:19:48: INFO[Main]:   ServerMap: load block [us] __________________ 2058x  166.595
2024-09-11 16:19:48: INFO[Main]:   Sky::render() [us] __________________________  639x  145.095
2024-09-11 16:19:48: INFO[Main]:   renderMap(): animated meshes [#] ____________  639x  69.184
2024-09-11 16:19:48: INFO[Main]:   renderMap(): transparent buffers [#] ________  639x  79.275
2024-09-11 16:19:48: INFO[Main]:   renderMap(SHADOW SOLID): draw meshes [ms] ___  625x   2.345
2024-09-11 16:19:48: INFO[Main]:   renderMap(SHADOW SOLID): drawcalls [#] ______  625x  219.633
2024-09-11 16:19:48: INFO[Main]:   renderMap(SHADOW SOLID): material swaps [#] _  625x   27.76
2024-09-11 16:19:48: INFO[Main]:   renderMap(SHADOW SOLID): vertices drawn [#] _  625x  51377.2
2024-09-11 16:19:48: INFO[Main]:   renderMap(SHADOW TRANS): draw meshes [ms] ___   79x   0.518
2024-09-11 16:19:48: INFO[Main]:   renderMap(SHADOW TRANS): drawcalls [#] ______   79x  72.177
2024-09-11 16:19:48: INFO[Main]:   renderMap(SHADOW TRANS): material swaps [#] _   79x  72.177
2024-09-11 16:19:48: INFO[Main]:   renderMap(SHADOW TRANS): vertices drawn [#] _   79x   44555
2024-09-11 16:19:48: INFO[Main]:   renderMap(SOLID): draw meshes [ms] __________  639x  12.076
2024-09-11 16:19:48: INFO[Main]:   renderMap(SOLID): drawcalls [#] _____________  639x  1363.85
2024-09-11 16:19:48: INFO[Main]:   renderMap(SOLID): material swaps [#] ________  639x  56.345
2024-09-11 16:19:48: INFO[Main]:   renderMap(SOLID): vertices drawn [#] ________  639x  362829
2024-09-11 16:19:48: INFO[Main]:   renderMap(TRANSPARENT): draw meshes [ms] ____  639x   0.754
2024-09-11 16:19:48: INFO[Main]:   renderMap(TRANSPARENT): drawcalls [#] _______  639x  79.275
2024-09-11 16:19:48: INFO[Main]:   renderMap(TRANSPARENT): material swaps [#] __  639x  79.275
2024-09-11 16:19:48: INFO[Main]:   renderMap(TRANSPARENT): vertices drawn [#] __  639x  50244.8
```

**Comment by appgurueu** [comment]
> I was able to record, but after running "inject", then "report" would throw a "data has no samples!" error. 
> Attaching the raw files for your examination: [...]

I looked at this in Hotspot, but it looks like the crucial `sched:sched_stat_sleep` event is missing (only `sched:sched_switch` and `sched:sched_process_exit` seem to be present) so I couldn't get much out of it.

I tried collecting this event myself, but it seems `perf` is not collecting sleep events for me either.

**Comment by sfan5** [comment]
> Not under load:
>
> Under load:

Honestly nothing really stands out here.
Sure, map loading is much slower but you tested with my PR included so it already doesn't affect the main thread.
But the server step duration barely changes:
```
2024-09-11 16:17:02: INFO[Main]:   Server::AsyncRunStep() [ms] _________________  276x   0.985
vs
2024-09-11 16:19:48: INFO[Main]:   Server::AsyncRunStep() [ms] _________________  865x   1.394
```
(note that the default time budget for one step is 11ms)

**Comment by Warr1024** [comment]
> And the server step duration barely changes

If that's what your instrumentation tells you, then your instrumentation is not very useful.  It looks like it's telling you the mean step time, which is consistent with what I observed.  Unfortunately, this problem isn't about mean step time, it's about the impact of large outliers.

If the game spends 10 seconds running 1000 globalsteps, and then spends 10 seconds running 1 globalstep, the average step size is ~ 0.019, which looks very reasonable on paper.  However, you know that half of the time, you have up to 10 seconds to wait for another globalstep, which disrupts a lot of universal-across-games things very badly, like entity physics and collison.  If you look at the average amount of time players are waiting for the next step, it's a little over 2.5, which gives you a better picture of the problem.  Step sizes need to be weighted by their duration.

---

Here's my refined test methodology with more rigor:

- Create a new `minetest_game` world.
- Install [szutilpack](https://content.minetest.net/packages/Warr1024/szutilpack/) and enable `szutil_lagometer`.
- Install the following custom mod into `worldmods`:

```lua
minetest.register_on_joinplayer(function(player)
		-- Move player along a set path with set timing to trigger mapblock loading
		local pname = player:get_player_name()
		local qty = 0
		local function set()
			player = minetest.get_player_by_name(pname)
			if not player then return end
			player:set_pos(vector.new(qty * 1024, 0, 0))
			player:set_look_horizontal(0)
			player:set_look_vertical(0)
			qty = qty + 1
			minetest.after(5, set)
		end
		set()

		-- Make sure the lagometer is enabled
		local meta = player:get_meta()
		meta:set_string("lagometer", "1")
		meta:set_string("lagometer_basic", "")

		-- Hide HUD elements and boost the contrast/readability of lagometer
		player:hud_add({
				type = "image",
				text = "[combine:1x1^[noalpha^[opacity:240",
				position = {x = 0.5, y = 0.5},
				direction = 0,
				scale = {x = -100, y = -100},
				offset = {x = 0, y = 0},
				z_index = -1000,
			})
		player:hud_set_flags({
				crosshair = false,
				wielditem = false,
				healthbar = false,
				breathbar = false,
				hotbar = false,
			})

		-- Show a formspec to prevent accidental camera turns
		minetest.after(0, function()
				player:set_formspec_prepend("")
				minetest.show_formspec(pname, "empty", "bgcolor[#00000000;false]")
			end)
	end)
```

- Make sure your view range is something exhorbitant like `viewing_range = 500`.
- Start the world in **singleplayer** mode, with **Host Server *unchecked***.
- Wait until the uptime in the bottom right shows 1:30.  The lagometer shows the previous 1 minute of statistics, so the extra :30 guards against "warm up" outliers (negligible in my tests, but just to be sure).
- To capture a sample, just take a screenshot at 1:30.
- Run a couple of cycles of this before starting any real measurements, to be certain the map has already been generated within the test areas.

---

Under no load, things are pretty much as you'd expect.  We're able to consistently run steps close to 60Hz, and the occasional excursion up to ~100ms doesn't harm gameplay very much.

**5.9.0:**
![noload-590](https://github.com/user-attachments/assets/319627a4-0706-4f77-96ec-88b402218379)

**5.10.0-dev with PR 15151:**
![noload-pr15151](https://github.com/user-attachments/assets/feb25c6b-5f7b-443b-b692-4dcfacff0843)

---

Now, under disk saturation, we can see that PR 15151 is a significant improvement over 5.9.0, but doesn't fix the problem entirely.  While the _mean_ times are comparable, the lag excursions in 5.9.0 are terrible, as high as nearly 10 seconds at one point, with a 3rd quartile above 1 second, so between 25% and 50% of the time, it's "unplayable".  With the patch, the max is down to only ~2.5 seconds, and the third quartile is barely elevated, so it's only "unplayable" between 0% and 25% of the time.

**5.9.0:**
![muchload-590](https://github.com/user-attachments/assets/7a1deeaf-197a-40c8-9002-68a495bc780c)

**5.10.0-dev with PR 15151:**
![muchload-pr15151](https://github.com/user-attachments/assets/4d064cd1-6e32-4b3e-87dd-0246eca5a577)

**Comment by sfan5** [comment]
> If that's what your instrumentation tells you, then your instrumentation is not very useful. It looks like it's telling you the mean step time, which is consistent with what I observed. Unfortunately, this problem isn't about mean step time, it's about the impact of large outliers.

It is the mean value, yes. Our profiler does not have support for quartiles.
